From f59ff33447ea8312ed8c4bf7e87c7d5409f0d2b9 Mon Sep 17 00:00:00 2001
From: Hugo Mano <hugo@zml.ai>
Date: Wed, 12 Feb 2025 13:10:04 +0100
Subject: [PATCH] Various macOS QOL enchancements

This PR adds various small quality of life improvements to macOS builds:

- drop the .so suffix for PjRt plugin targets (.dylib on macOS)
- add compatibility with Apple Command Line Tools (no need for Xcode anymore)
only export the GetPjrtApi symbol on macOS
- leverage bazel's cc_binary.additional_linker_inputs instead of using deps

PR: https://github.com/openxla/xla/pull/16696

Co-authored-by: Steeve Morin <steeve@zml.ai>
---
 xla/pjrt/c/BUILD | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/xla/pjrt/c/BUILD b/xla/pjrt/c/BUILD
index a0485b6a43..6f67ee6b78 100644
--- a/xla/pjrt/c/BUILD
+++ b/xla/pjrt/c/BUILD
@@ -321,9 +321,14 @@ cc_library(
 
 # PJRT CPU plugin.
 xla_cc_binary(
-    name = "pjrt_c_api_cpu_plugin.so",
+    name = "pjrt_c_api_cpu_plugin",
+    additional_linker_inputs = if_macos([], [":pjrt_c_api_cpu_version_script.lds"]),
     linkopts = if_macos(
-        [],
+        [
+            "-Wl,-exported_symbol,_GetPjrtApi",
+            "-Wl,-install_name,@rpath/pjrt_c_api_cpu_plugin.dylib",
+            "-Wl,-undefined,error",
+        ],
         [
             "-Wl,--version-script,$(location :pjrt_c_api_cpu_version_script.lds)",
             "-Wl,--no-undefined",
@@ -336,10 +341,7 @@ xla_cc_binary(
         "notsan",
     ],
     visibility = ["//visibility:public"],
-    deps = [
-        ":pjrt_c_api_cpu",
-        ":pjrt_c_api_cpu_version_script.lds",
-    ],
+    deps = [":pjrt_c_api_cpu"],
 )
 
 cc_library(
@@ -408,7 +410,8 @@ cc_library(
 
 # PJRT GPU plugin. Can be configured to be built for CUDA or ROCM.
 xla_cc_binary(
-    name = "pjrt_c_api_gpu_plugin.so",
+    name = "pjrt_c_api_gpu_plugin",
+    additional_linker_inputs = [":pjrt_c_api_gpu_version_script.lds"],
     linkopts = [
         "-Wl,--version-script,$(location :pjrt_c_api_gpu_version_script.lds)",
         "-Wl,--no-undefined",
@@ -422,7 +425,6 @@ xla_cc_binary(
     ],
     deps = [
         ":pjrt_c_api_gpu",
-        ":pjrt_c_api_gpu_version_script.lds",
         "//xla/service:gpu_plugin",
     ] + if_cuda_is_configured([
         "//xla/stream_executor:cuda_platform",
-- 
2.39.5 (Apple Git-154)
